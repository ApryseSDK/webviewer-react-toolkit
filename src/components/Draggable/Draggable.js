'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
exports.Draggable = void 0;
var tslib_1 = require('tslib');
var classnames_1 = tslib_1.__importDefault(require('classnames'));
var react_1 = tslib_1.__importStar(require('react'));
var react_dnd_1 = require('react-dnd');
var react_dnd_html5_backend_1 = require('react-dnd-html5-backend');
var react_motion_1 = require('react-motion');
var hooks_1 = require('../../hooks');
var utils_1 = require('../../utils');
var ItemTypes = { Draggable: 'draggable' };
// Quick animation with no "bounce".
var SPRING = { stiffness: 300, damping: 30 };
exports.Draggable = (0, react_1.forwardRef)(function(_a, ref) {
  var index = _a.index,
    disableDrag = _a.disableDrag,
    hideDragPreview = _a.hideDragPreview,
    preventAnimation = _a.preventAnimation,
    onRenderChildren = _a.onRenderChildren,
    onMove = _a.onMove,
    onDragChange = _a.onDragChange,
    children = _a.children,
    className = _a.className,
    divProps = tslib_1.__rest(_a, [
      'index',
      'disableDrag',
      'hideDragPreview',
      'preventAnimation',
      'onRenderChildren',
      'onMove',
      'onDragChange',
      'children',
      'className',
    ]);
  var draggableRef = (0, react_1.useRef)(null);
  (0, react_1.useImperativeHandle)(ref, function() {
    return draggableRef.current;
  });
  /* --- Drag and drop settings. --- */
  var _b = (0, react_dnd_1.useDrop)({
      accept: ItemTypes.Draggable,
      hover: function(dragItem) {
        // Previous index.
        var fromIndex = dragItem.index;
        // Index it has been dragged to.
        var toIndex = index;
        // Cancel if index has not changed.
        if (fromIndex === toIndex) return;
        // Call onMove when index changes.
        var success =
          onMove === null || onMove === void 0
            ? void 0
            : onMove(fromIndex, toIndex);
        // Set the item index to be the new index.
        if (success) dragItem.index = toIndex;
      },
    }),
    drop = _b[1];
  var _c = (0, react_dnd_1.useDrag)({
      item: { type: ItemTypes.Draggable, index: index },
      collect: function(monitor) {
        return { isDragging: monitor.isDragging() };
      },
      canDrag: !disableDrag,
    }),
    isDragging = _c[0].isDragging,
    drag = _c[1],
    preview = _c[2];
  (0, react_1.useEffect)(
    function() {
      if (hideDragPreview)
        preview((0, react_dnd_html5_backend_1.getEmptyImage)(), {
          captureDraggingState: true,
        });
    },
    [hideDragPreview, preview],
  );
  // Call onDragChange whenever isDragging changes.
  var onDragChangeRef = (0, hooks_1.useCurrentRef)(onDragChange);
  (0, react_1.useEffect)(
    function() {
      var _a;
      if (isDragging && onDragChangeRef.current) {
        (_a = onDragChangeRef.current) === null || _a === void 0
          ? void 0
          : _a.call(onDragChangeRef, true);
        return function() {
          var _a;
          return (_a = onDragChangeRef.current) === null || _a === void 0
            ? void 0
            : _a.call(onDragChangeRef, false);
        }; // eslint-disable-line react-hooks/exhaustive-deps
      }
      return;
    },
    [isDragging, onDragChangeRef],
  );
  drag(drop(draggableRef));
  /* --- Animation settings. --- */
  var _d = (0, react_1.useState)({ x: 0, y: 0 }),
    coords = _d[0],
    setCoords = _d[1];
  var prevIndex = (0, react_1.useRef)(index);
  var noAnimation = (0, hooks_1.useCurrentRef)(isDragging || preventAnimation);
  // Find the difference in position between new index and previous index. Then,
  // get the difference in position and set that as the starting point for the
  // animation.
  (0, react_1.useEffect)(
    function() {
      var _a;
      // Return early if no animation, since we do not want to compute the
      // previous location.
      if (!draggableRef.current || noAnimation.current) return;
      // Get sibling that occupies previous spot to find params.
      var indexDiff = prevIndex.current - index;
      var prev = (0, utils_1.getSibling)(draggableRef.current, indexDiff);
      // Get the coordinates of the previous item.
      var _b =
          (_a =
            prev === null || prev === void 0
              ? void 0
              : prev.getBoundingClientRect()) !== null && _a !== void 0
            ? _a
            : {},
        prevLeft = _b.left,
        prevTop = _b.top;
      // Get the coordinates of the current item.
      var _c = draggableRef.current.getBoundingClientRect(),
        left = _c.left,
        top = _c.top;
      // Get the deltas.
      var deltaX = prevLeft === undefined ? 0 : prevLeft - left;
      var deltaY = prevTop === undefined ? 0 : prevTop - top;
      // Set the coordinates to the distance.
      setCoords({ x: deltaX / 6, y: deltaY / 6 });
      // Store index for next swap.
      prevIndex.current = index;
    },
    [index, noAnimation],
  );
  // Whenever coords change, revert back to zero.
  (0, react_1.useEffect)(
    function() {
      if (coords.x === 0 && coords.y === 0) return;
      requestAnimationFrame(function() {
        return setCoords({ x: 0, y: 0 });
      });
    },
    [coords],
  );
  var motionStyle = (0, react_1.useMemo)(
    function() {
      if (noAnimation.current) return { x: 0, y: 0 };
      // Only spring when returning to zero. This lets it "snap" to the previous
      // location, then "spring" back to the new location.
      return {
        x:
          coords.x === 0
            ? (0, react_motion_1.spring)(coords.x, SPRING)
            : coords.x,
        y:
          coords.y === 0
            ? (0, react_motion_1.spring)(coords.y, SPRING)
            : coords.y,
      };
    },
    [coords, noAnimation],
  );
  var draggableClass = (0, classnames_1.default)(
    'ui__base ui__draggable',
    className,
  );
  var onMotionRender = (0, react_1.useCallback)(
    function(_a) {
      var x = _a.x,
        y = _a.y;
      var inMotion = !!(x || y);
      return react_1.default.createElement(
        'div',
        tslib_1.__assign({}, divProps, {
          ref: draggableRef,
          className: draggableClass,
        }),
        react_1.default.createElement(
          'div',
          {
            style: {
              WebkitTransform: 'translate3d('
                .concat(x, 'px, ')
                .concat(y, 'px, 0)'),
              transform: 'translate3d('.concat(x, 'px, ').concat(y, 'px, 0)'),
            },
            className: (0, classnames_1.default)('ui__draggable__animated', {
              'ui__draggable__animated--inMotion': inMotion,
            }),
          },
          onRenderChildren ? onRenderChildren(isDragging) : children,
        ),
      );
    },
    [children, divProps, draggableClass, isDragging, onRenderChildren],
  );
  return react_1.default.createElement(
    react_motion_1.Motion,
    { style: motionStyle },
    onMotionRender,
  );
});
//# sourceMappingURL=Draggable.js.map
