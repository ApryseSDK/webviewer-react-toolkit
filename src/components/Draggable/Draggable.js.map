{"version":3,"file":"Draggable.js","sourceRoot":"","sources":["Draggable.tsx"],"names":[],"mappings":";;;;AAAA,kEAAoC;AACpC,qDAUe;AACf,uCAAiE;AACjE,mEAAwD;AACxD,6CAA8E;AAC9E,qCAA4C;AAC5C,qCAAyC;AAEzC,IAAM,SAAS,GAAG,EAAE,SAAS,EAAE,WAAW,EAAE,CAAC;AA6C7C,oCAAoC;AACpC,IAAM,MAAM,GAAuB,EAAE,SAAS,EAAE,GAAG,EAAE,OAAO,EAAE,EAAE,EAAE,CAAC;AAEtD,QAAA,SAAS,GAAG,IAAA,kBAAU,EACjC,UACE,EAWC,EACD,GAAG;IAXD,IAAA,KAAK,WAAA,EACL,WAAW,iBAAA,EACX,eAAe,qBAAA,EACf,gBAAgB,sBAAA,EAChB,gBAAgB,sBAAA,EAChB,MAAM,YAAA,EACN,YAAY,kBAAA,EACZ,QAAQ,cAAA,EACR,SAAS,eAAA,EACN,QAAQ,sBAVb,sIAWC,CADY;IAIb,IAAM,YAAY,GAAG,IAAA,cAAM,EAAiB,IAAI,CAAC,CAAC;IAClD,IAAA,2BAAmB,EAAC,GAAG,EAAE,cAAM,OAAA,YAAY,CAAC,OAAyB,EAAtC,CAAsC,CAAC,CAAC;IAEvE,qCAAqC;IAE/B,IAAA,KAAW,IAAA,mBAAO,EAAC;QACvB,MAAM,EAAE,SAAS,CAAC,SAAS;QAC3B,KAAK,YAAC,QAAkB;YACtB,kBAAkB;YAClB,IAAM,SAAS,GAAG,QAAQ,CAAC,KAAK,CAAC;YAEjC,gCAAgC;YAChC,IAAM,OAAO,GAAG,KAAK,CAAC;YAEtB,mCAAmC;YACnC,IAAI,SAAS,KAAK,OAAO;gBAAE,OAAO;YAElC,kCAAkC;YAClC,IAAM,OAAO,GAAG,MAAM,aAAN,MAAM,uBAAN,MAAM,CAAG,SAAS,EAAE,OAAO,CAAC,CAAC;YAE7C,0CAA0C;YAC1C,IAAI,OAAO;gBAAE,QAAQ,CAAC,KAAK,GAAG,OAAO,CAAC;QACxC,CAAC;KACF,CAAC,EAlBO,IAAI,QAkBX,CAAC;IAEG,IAAA,KAAkC,IAAA,mBAAO,EAAC;QAC9C,IAAI,EAAE,EAAE,IAAI,EAAE,SAAS,CAAC,SAAS,EAAE,KAAK,OAAA,EAAE;QAC1C,OAAO,EAAE,UAAC,OAAO,IAAK,OAAA,CAAC,EAAE,UAAU,EAAE,OAAO,CAAC,UAAU,EAAE,EAAE,CAAC,EAAtC,CAAsC;QAC5D,OAAO,EAAE,CAAC,WAAW;KACtB,CAAC,EAJO,UAAU,mBAAA,EAAI,IAAI,QAAA,EAAE,OAAO,QAIlC,CAAC;IAEH,IAAA,iBAAS,EAAC;QACR,IAAI,eAAe;YAAE,OAAO,CAAC,IAAA,uCAAa,GAAE,EAAE,EAAE,oBAAoB,EAAE,IAAI,EAAE,CAAC,CAAC;IAChF,CAAC,EAAE,CAAC,eAAe,EAAE,OAAO,CAAC,CAAC,CAAC;IAE/B,iDAAiD;IACjD,IAAM,eAAe,GAAG,IAAA,qBAAa,EAAC,YAAY,CAAC,CAAC;IACpD,IAAA,iBAAS,EAAC;;QACR,IAAI,UAAU,IAAI,eAAe,CAAC,OAAO,EAAE;YACzC,MAAA,eAAe,CAAC,OAAO,gEAAG,IAAI,CAAC,CAAC;YAChC,OAAO,sBAAM,OAAA,MAAA,eAAe,CAAC,OAAO,gEAAG,KAAK,CAAC,CAAA,EAAA,CAAC,CAAC,kDAAkD;SAClG;QACD,OAAO;IACT,CAAC,EAAE,CAAC,UAAU,EAAE,eAAe,CAAC,CAAC,CAAC;IAElC,IAAI,CAAC,IAAI,CAAC,YAAY,CAAC,CAAC,CAAC;IAEzB,iCAAiC;IAE3B,IAAA,KAAsB,IAAA,gBAAQ,EAAC,EAAE,CAAC,EAAE,CAAC,EAAE,CAAC,EAAE,CAAC,EAAE,CAAC,EAA7C,MAAM,QAAA,EAAE,SAAS,QAA4B,CAAC;IACrD,IAAM,SAAS,GAAG,IAAA,cAAM,EAAC,KAAK,CAAC,CAAC;IAEhC,IAAM,WAAW,GAAG,IAAA,qBAAa,EAAC,UAAU,IAAI,gBAAgB,CAAC,CAAC;IAElE,8EAA8E;IAC9E,4EAA4E;IAC5E,aAAa;IACb,IAAA,iBAAS,EAAC;;QACR,oEAAoE;QACpE,qBAAqB;QACrB,IAAI,CAAC,YAAY,CAAC,OAAO,IAAI,WAAW,CAAC,OAAO;YAAE,OAAO;QAEzD,0DAA0D;QAC1D,IAAM,SAAS,GAAG,SAAS,CAAC,OAAO,GAAG,KAAK,CAAC;QAC5C,IAAM,IAAI,GAAG,IAAA,kBAAU,EAAC,YAAY,CAAC,OAAO,EAAE,SAAS,CAAC,CAAC;QAEzD,4CAA4C;QACtC,IAAA,KAAmC,MAAA,IAAI,aAAJ,IAAI,uBAAJ,IAAI,CAAE,qBAAqB,EAAE,mCAAI,EAAE,EAA9D,QAAQ,UAAA,EAAO,OAAO,SAAwC,CAAC;QAE7E,2CAA2C;QACrC,IAAA,KAAgB,YAAY,CAAC,OAAO,CAAC,qBAAqB,EAAE,EAA1D,IAAI,UAAA,EAAE,GAAG,SAAiD,CAAC;QAEnE,kBAAkB;QAClB,IAAM,MAAM,GAAG,QAAQ,KAAK,SAAS,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,QAAQ,GAAG,IAAI,CAAC;QAC5D,IAAM,MAAM,GAAG,OAAO,KAAK,SAAS,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,OAAO,GAAG,GAAG,CAAC;QAEzD,uCAAuC;QACvC,SAAS,CAAC,EAAE,CAAC,EAAE,MAAM,GAAG,CAAC,EAAE,CAAC,EAAE,MAAM,GAAG,CAAC,EAAE,CAAC,CAAC;QAE5C,6BAA6B;QAC7B,SAAS,CAAC,OAAO,GAAG,KAAK,CAAC;IAC5B,CAAC,EAAE,CAAC,KAAK,EAAE,WAAW,CAAC,CAAC,CAAC;IAEzB,+CAA+C;IAC/C,IAAA,iBAAS,EAAC;QACR,IAAI,MAAM,CAAC,CAAC,KAAK,CAAC,IAAI,MAAM,CAAC,CAAC,KAAK,CAAC;YAAE,OAAO;QAC7C,qBAAqB,CAAC,cAAM,OAAA,SAAS,CAAC,EAAE,CAAC,EAAE,CAAC,EAAE,CAAC,EAAE,CAAC,EAAE,CAAC,EAAzB,CAAyB,CAAC,CAAC;IACzD,CAAC,EAAE,CAAC,MAAM,CAAC,CAAC,CAAC;IAEb,IAAM,WAAW,GAAG,IAAA,eAAO,EAAC;QAC1B,IAAI,WAAW,CAAC,OAAO;YAAE,OAAO,EAAE,CAAC,EAAE,CAAC,EAAE,CAAC,EAAE,CAAC,EAAE,CAAC;QAC/C,0EAA0E;QAC1E,oDAAoD;QACpD,OAAO;YACL,CAAC,EAAE,MAAM,CAAC,CAAC,KAAK,CAAC,CAAC,CAAC,CAAC,IAAA,qBAAM,EAAC,MAAM,CAAC,CAAC,EAAE,MAAM,CAAC,CAAC,CAAC,CAAC,MAAM,CAAC,CAAC;YACvD,CAAC,EAAE,MAAM,CAAC,CAAC,KAAK,CAAC,CAAC,CAAC,CAAC,IAAA,qBAAM,EAAC,MAAM,CAAC,CAAC,EAAE,MAAM,CAAC,CAAC,CAAC,CAAC,MAAM,CAAC,CAAC;SACxD,CAAC;IACJ,CAAC,EAAE,CAAC,MAAM,EAAE,WAAW,CAAC,CAAC,CAAC;IAE1B,IAAM,cAAc,GAAG,IAAA,oBAAU,EAAC,wBAAwB,EAAE,SAAS,CAAC,CAAC;IAEvE,IAAM,cAAc,GAAG,IAAA,mBAAW,EAChC,UAAC,EAAoB;YAAlB,CAAC,OAAA,EAAE,CAAC,OAAA;QACL,IAAM,QAAQ,GAAG,CAAC,CAAC,CAAC,CAAC,IAAI,CAAC,CAAC,CAAC;QAE5B,OAAO,CACL,0DAAS,QAAQ,IAAE,GAAG,EAAE,YAAY,EAAE,SAAS,EAAE,cAAc;YAC7D,uCACE,KAAK,EAAE;oBACL,eAAe,EAAE,sBAAe,CAAC,iBAAO,CAAC,WAAQ;oBACjD,SAAS,EAAE,sBAAe,CAAC,iBAAO,CAAC,WAAQ;iBAC5C,EACD,SAAS,EAAE,IAAA,oBAAU,EAAC,yBAAyB,EAAE;oBAC/C,mCAAmC,EAAE,QAAQ;iBAC9C,CAAC,IAED,gBAAgB,CAAC,CAAC,CAAC,gBAAgB,CAAC,UAAU,CAAC,CAAC,CAAC,CAAC,QAAQ,CACvD,CACF,CACP,CAAC;IACJ,CAAC,EACD,CAAC,QAAQ,EAAE,QAAQ,EAAE,cAAc,EAAE,UAAU,EAAE,gBAAgB,CAAC,CACnE,CAAC;IAEF,OAAO,8BAAC,qBAAM,IAAC,KAAK,EAAE,WAAW,IAAG,cAAc,CAAU,CAAC;AAC/D,CAAC,CACF,CAAC"}