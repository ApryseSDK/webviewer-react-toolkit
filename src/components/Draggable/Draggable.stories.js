'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
exports.WithOnRenderChildren = exports.Basic = void 0;
var tslib_1 = require('tslib');
var addon_knobs_1 = require('@storybook/addon-knobs');
var react_1 = tslib_1.__importDefault(require('react'));
var action_1 = require('../../storybook-helpers/action/action');
var DndMultiProvider_1 = require('../DndMultiProvider');
var Draggable_1 = require('../Draggable');
var README_md_1 = tslib_1.__importDefault(require('./README.md'));
exports.default = {
  title: 'Components/Draggable',
  component: Draggable_1.Draggable,
  parameters: { readme: README_md_1.default },
};
var commonStyle = {
  padding: 16,
  background: 'lightgray',
  color: 'black',
  width: 200,
  height: 50,
  textAlign: 'center',
};
var Basic = function() {
  return react_1.default.createElement(
    DndMultiProvider_1.DndMultiProvider,
    null,
    react_1.default.createElement(
      Draggable_1.Draggable,
      {
        index: 0,
        onDragChange: (0, action_1.action)('onDragChange'),
        disableDrag: (0, addon_knobs_1.boolean)('disableDrag', false),
        hideDragPreview: (0, addon_knobs_1.boolean)('hideDragPreview', false),
      },
      react_1.default.createElement(
        'div',
        {
          style: tslib_1.__assign(tslib_1.__assign({}, commonStyle), {
            border: '1px solid red',
          }),
        },
        'This div is draggable!',
      ),
    ),
  );
};
exports.Basic = Basic;
var WithOnRenderChildren = function() {
  return react_1.default.createElement(
    DndMultiProvider_1.DndMultiProvider,
    null,
    react_1.default.createElement(Draggable_1.Draggable, {
      index: 0,
      onRenderChildren: function(isDragging) {
        return react_1.default.createElement(
          'div',
          {
            style: tslib_1.__assign(tslib_1.__assign({}, commonStyle), {
              border: '1px solid '.concat(isDragging ? 'blue' : 'red'),
            }),
          },
          isDragging ? 'This div is being dragged!' : 'This div is draggable!',
        );
      },
      onDragChange: (0, action_1.action)('onDragChange'),
      disableDrag: (0, addon_knobs_1.boolean)('disableDrag', false),
      hideDragPreview: (0, addon_knobs_1.boolean)('hideDragPreview', false),
    }),
  );
};
exports.WithOnRenderChildren = WithOnRenderChildren;
//# sourceMappingURL=Draggable.stories.js.map
