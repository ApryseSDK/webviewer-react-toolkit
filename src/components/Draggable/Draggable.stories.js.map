{"version":3,"file":"Draggable.stories.js","sourceRoot":"","sources":["Draggable.stories.tsx"],"names":[],"mappings":";;;;AAAA,sDAAiD;AACjD,wDAA6C;AAC7C,gEAA+D;AAC/D,wDAAuD;AACvD,0CAAyC;AACzC,kEAAiC;AAEjC,kBAAe,EAAE,KAAK,EAAE,sBAAsB,EAAE,SAAS,EAAE,qBAAS,EAAE,UAAU,EAAE,EAAE,MAAM,qBAAA,EAAE,EAAE,CAAC;AAE/F,IAAM,WAAW,GAAkB;IACjC,OAAO,EAAE,EAAE;IACX,UAAU,EAAE,WAAW;IACvB,KAAK,EAAE,OAAO;IACd,KAAK,EAAE,GAAG;IACV,MAAM,EAAE,EAAE;IACV,SAAS,EAAE,QAAQ;CACpB,CAAC;AAEK,IAAM,KAAK,GAAG,cAAM,OAAA,CACzB,8BAAC,mCAAgB;IACf,8BAAC,qBAAS,IACR,KAAK,EAAE,CAAC,EACR,YAAY,EAAE,IAAA,eAAM,EAAC,cAAc,CAAC,EACpC,WAAW,EAAE,IAAA,qBAAO,EAAC,aAAa,EAAE,KAAK,CAAC,EAC1C,eAAe,EAAE,IAAA,qBAAO,EAAC,iBAAiB,EAAE,KAAK,CAAC;QAElD,uCAAK,KAAK,wCAAO,WAAW,KAAE,MAAM,EAAE,eAAe,gCAAgC,CAC3E,CACK,CACpB,EAX0B,CAW1B,CAAC;AAXW,QAAA,KAAK,SAWhB;AAEK,IAAM,oBAAoB,GAAG,cAAM,OAAA,CACxC,8BAAC,mCAAgB;IACf,8BAAC,qBAAS,IACR,KAAK,EAAE,CAAC,EACR,gBAAgB,EAAE,UAAC,UAAU,IAAK,OAAA,CAChC,uCAAK,KAAK,wCAAO,WAAW,KAAE,MAAM,EAAE,oBAAa,UAAU,CAAC,CAAC,CAAC,MAAM,CAAC,CAAC,CAAC,KAAK,CAAE,OAC7E,UAAU,CAAC,CAAC,CAAC,4BAA4B,CAAC,CAAC,CAAC,wBAAwB,CACjE,CACP,EAJiC,CAIjC,EACD,YAAY,EAAE,IAAA,eAAM,EAAC,cAAc,CAAC,EACpC,WAAW,EAAE,IAAA,qBAAO,EAAC,aAAa,EAAE,KAAK,CAAC,EAC1C,eAAe,EAAE,IAAA,qBAAO,EAAC,iBAAiB,EAAE,KAAK,CAAC,GAClD,CACe,CACpB,EAdyC,CAczC,CAAC;AAdW,QAAA,oBAAoB,wBAc/B"}