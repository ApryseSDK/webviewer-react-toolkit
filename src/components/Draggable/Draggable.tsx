import classnames from 'classnames';
import React, {
  forwardRef,
  HTMLAttributes,
  ReactNode,
  useCallback,
  useEffect,
  useImperativeHandle,
  useMemo,
  useRef,
  useState,
} from 'react';
import { DragObjectWithType, useDrag, useDrop } from 'react-dnd';
import { getEmptyImage } from 'react-dnd-html5-backend';
import { Motion, PlainStyle, spring, SpringHelperConfig } from 'react-motion';
import { useCurrentRef } from '../../hooks';
import { getSibling } from '../../utils';

const ItemTypes = { Draggable: 'draggable' };

export interface DraggableProps extends HTMLAttributes<HTMLDivElement> {
  /**
   * The current index of the draggable wrapper.
   */
  index: number;
  /**
   * Prevent this draggable wrapper from dragging.
   */
  disableDrag?: boolean;
  /**
   * Hides the snapshot preview of the item while dragging. The most common use
   * case would be if you're implementing a custom drag layer and you don't want
   * the preview snapshot to clash with it.
   */
  hideDragPreview?: boolean;
  /**
   * Prevent the move-to-location animation of this draggable.
   */
  preventAnimation?: boolean;
  /**
   * Call instead of providing children if you wish to use the `isDragging`
   * prop.
   * @param isDragging Is the child dragging.
   */
  onRenderChildren?(isDragging: boolean): ReactNode;
  /**
   * If given, will be called any time the draggable wrapper is moved. Returns
   * whether the move was successful.
   * @param fromIndex The previous index of the item.
   * @param toIndex The next index of the item.
   */
  onMove?(fromIndex: number, toIndex: number): boolean;
  /**
   * Called whenever the dnd property `isDragging` changes.
   * @param isDragging Is the child dragging.
   */
  onDragChange?(isDragging: boolean): void;
}

interface DragItem extends DragObjectWithType {
  index: number;
}

// Quick animation with no "bounce".
const SPRING: SpringHelperConfig = { stiffness: 300, damping: 30 };

export const Draggable = forwardRef<HTMLDivElement, DraggableProps>(
  (
    {
      index,
      disableDrag,
      hideDragPreview,
      preventAnimation,
      onRenderChildren,
      onMove,
      onDragChange,
      children,
      className,
      ...divProps
    },
    ref,
  ) => {
    const draggableRef = useRef<HTMLDivElement>(null);
    useImperativeHandle(ref, () => draggableRef.current!);

    /* --- Drag and drop settings. --- */

    const [, drop] = useDrop({
      accept: ItemTypes.Draggable,
      hover(dragItem: DragItem) {
        // Previous index.
        const fromIndex = dragItem.index;

        // Index it has been dragged to.
        const toIndex = index;

        // Cancel if index has not changed.
        if (fromIndex === toIndex) return;

        // Call onMove when index changes.
        const success = onMove?.(fromIndex, toIndex);

        // Set the item index to be the new index.
        if (success) dragItem.index = toIndex;
      },
    });

    const [{ isDragging }, drag, preview] = useDrag({
      item: { type: ItemTypes.Draggable, index },
      collect: monitor => ({ isDragging: monitor.isDragging() }),
      canDrag: !disableDrag,
    });

    useEffect(() => {
      if (hideDragPreview) preview(getEmptyImage(), { captureDraggingState: true });
    }, [hideDragPreview, preview]);

    // Call onDragChange whenever isDragging changes.
    const onDragChangeRef = useCurrentRef(onDragChange);
    useEffect(() => {
      if (isDragging && onDragChangeRef.current) {
        onDragChangeRef.current?.(true);
        return () => onDragChangeRef.current?.(false); // eslint-disable-line react-hooks/exhaustive-deps
      }
      return;
    }, [isDragging, onDragChangeRef]);

    drag(drop(draggableRef));

    /* --- Animation settings. --- */

    const [coords, setCoords] = useState({ x: 0, y: 0 });
    const prevIndex = useRef(index);

    // Find the difference in position between new index and previous index. Then,
    // get the difference in position and set that as the starting point for the
    // animation.
    useEffect(() => {
      if (!draggableRef.current) return;

      // Get sibling that occupies previous spot to find params.
      const indexDiff = prevIndex.current - index;
      const prev = getSibling(draggableRef.current, indexDiff);

      // Get the coordinates of the previous item.
      const { left: prevLeft, top: prevTop } = prev?.getBoundingClientRect() ?? {};

      // Get the coordinates of the current item.
      const { left, top } = draggableRef.current.getBoundingClientRect();

      // Get the deltas.
      const deltaX = prevLeft === undefined ? 0 : prevLeft - left;
      const deltaY = prevTop === undefined ? 0 : prevTop - top;

      // Set the coordinates to the distance.
      setCoords({ x: deltaX / 6, y: deltaY / 6 });

      // Store index for next swap.
      prevIndex.current = index;
    }, [index]);

    // Whenever coords change, revert back to zero.
    useEffect(() => {
      if (coords.x === 0 && coords.y === 0) return;
      requestAnimationFrame(() => setCoords({ x: 0, y: 0 }));
    }, [coords]);

    // Only animate if returning back to 0. Otherwise, snap to translate so that
    // the animation looks like it's moving from it's previous location.
    const motionStyle = useMemo(
      () => ({
        x: coords.x === 0 && !isDragging && !preventAnimation ? spring(coords.x, SPRING) : coords.x,
        y: coords.y === 0 && !isDragging && !preventAnimation ? spring(coords.y, SPRING) : coords.y,
      }),
      [coords.x, coords.y, preventAnimation, isDragging],
    );

    const draggableClass = classnames('ui__base ui__draggable', className);

    const onMotionRender = useCallback(
      ({ x, y }: PlainStyle) => {
        const inMotion = !!(x || y);

        return (
          <div {...divProps} ref={draggableRef} className={draggableClass}>
            <div
              style={{
                WebkitTransform: `translate3d(${x}px, ${y}px, 0)`,
                transform: `translate3d(${x}px, ${y}px, 0)`,
              }}
              className={classnames('ui__draggable__animated', {
                'ui__draggable__animated--inMotion': inMotion,
              })}
            >
              {onRenderChildren ? onRenderChildren(isDragging) : children}
            </div>
          </div>
        );
      },
      [children, divProps, draggableClass, isDragging, onRenderChildren],
    );

    return <Motion style={motionStyle}>{onMotionRender}</Motion>;
  },
);
