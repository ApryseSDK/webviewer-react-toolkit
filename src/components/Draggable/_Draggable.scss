.ui__draggable {
  display: inline-block;
}
