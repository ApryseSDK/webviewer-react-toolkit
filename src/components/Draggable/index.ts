export { Draggable as default, DraggableProps } from './Draggable';
