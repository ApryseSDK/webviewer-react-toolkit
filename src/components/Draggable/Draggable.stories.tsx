import { action } from '@storybook/addon-actions';
import { boolean } from '@storybook/addon-knobs';
import React, { CSSProperties } from 'react';
import { DndMultiProvider } from '../DndMultiProvider';
import { Draggable } from '../Draggable';
import docs from './README.md';

export default { title: 'Draggable', component: Draggable, parameters: { info: docs } };

const commonStyle: CSSProperties = {
  padding: 16,
  background: 'lightgray',
  width: 200,
  height: 50,
  textAlign: 'center',
};

export const Basic = () => (
  <DndMultiProvider>
    <Draggable
      index={0}
      onMove={action('onMove')}
      onDragChange={action('onDragChange')}
      disableDrag={boolean('disableDrag', false)}
      hideDragPreview={boolean('hideDragPreview', false)}
    >
      <div style={{ ...commonStyle, border: '1px solid red' }}>This div is draggable!</div>
    </Draggable>
  </DndMultiProvider>
);

export const WithOnRenderChildren = () => (
  <DndMultiProvider>
    <Draggable
      index={0}
      onRenderChildren={isDragging => (
        <div style={{ ...commonStyle, border: `1px solid ${isDragging ? 'blue' : 'red'}` }}>
          {isDragging ? 'This div is being dragged!' : 'This div is draggable!'}
        </div>
      )}
      onMove={action('onMove')}
      onDragChange={action('onDragChange')}
      disableDrag={boolean('disableDrag', false)}
      hideDragPreview={boolean('hideDragPreview', false)}
    />
  </DndMultiProvider>
);
