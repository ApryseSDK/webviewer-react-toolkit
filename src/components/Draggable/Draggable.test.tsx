import { mount } from 'enzyme';
import React from 'react';
import Draggable from '../Draggable';
import { DndProvider } from 'react-dnd';
import Backend from 'react-dnd-html5-backend';

describe('Draggable component', () => {
  it('renders its contents', () => {
    const draggable = mount(
      <DndProvider backend={Backend}>
        <Draggable index={0}>child</Draggable>
      </DndProvider>,
    );
    expect(draggable.find('.ui__draggable').length).toEqual(1);
  });

  it('snapshot renders default draggable', () => {
    const draggable = mount(
      <DndProvider backend={Backend}>
        <Draggable index={0}>child</Draggable>
      </DndProvider>,
    );
    expect(draggable).toMatchSnapshot();
  });
});
