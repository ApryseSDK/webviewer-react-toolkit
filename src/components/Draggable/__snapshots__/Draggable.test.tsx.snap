// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Draggable component snapshot renders default draggable 1`] = `
<DndProvider
  backend={[Function]}
>
  <Component
    index={0}
  >
    <Motion
      style={
        Object {
          "x": Object {
            "damping": 30,
            "precision": 0.01,
            "stiffness": 300,
            "val": 0,
          },
          "y": Object {
            "damping": 30,
            "precision": 0.01,
            "stiffness": 300,
            "val": 0,
          },
        }
      }
    >
      <div
        className="ui__base ui__draggable"
        style={
          Object {
            "WebkitTransform": "translate3d(0px, 0px, 0)",
            "transform": "translate3d(0px, 0px, 0)",
            "zIndex": undefined,
          }
        }
      >
        child
      </div>
    </Motion>
  </Component>
</DndProvider>
`;
